Problem batch

nwestallen opened the pull request:
Problems for review

- [ ] add_two_numbers https://leetcode.com/problems/add-two-numbers/
- [ ] longest_substring https://leetcode.com/problems/longest-substring-without-repeating-characters/
- [ ] median_two_arrays https://leetcode.com/problems/median-of-two-sorted-arrays/

nwestallen committed:
add_two_numbers

nwestallen committed:
longest substring

nwestallen committed:
median of two sorted arrays

Jaya-Hoyt commented on problems/add_two_numbers.py:
```diff
@@ -0,0 +1,41 @@
+        l3 = ListNode()
+        cur1 = l1
+        cur2 = l2
+        cur3 = l3
```
Using a third linked list is smart to preserve the data of the input. worth pointing out in an interview that you're trading off O(1) space to get a more predictable program

Jaya-Hoyt commented on problems/add_two_numbers.py:
```diff
@@ -0,0 +1,41 @@
+#https://leetcode.com/problems/add-two-numbers/submissions/
+
```
nit: put the leetcode problem number in your filename. in the beginning makes things easier to find as you start building up a library. e.g. `2_add_two_numbers.py`

Jaya-Hoyt commented on problems/longest_substring_w-o_repeats.py:
```diff
@@ -0,0 +1,32 @@
+#https://leetcode.com/submissions/detail/508592080/
```
I can't get to this page. maybe i need an active membership?

Jaya-Hoyt commented on problems/longest_substring_w-o_repeats.py:
```diff
@@ -0,0 +1,32 @@
+#https://leetcode.com/submissions/detail/508592080/
+
+class Solution(object):
```
Add your big O time and space analysis in a top comment to your functions

Jaya-Hoyt commented on problems/longest_substring_w-o_repeats.py:
```diff
@@ -0,0 +1,32 @@
+                #print(i, j, word)
+                if s[j] in word:
+                    #print(s[j])
+                    if len(word) > maxlen:
```
nit: `maxlen = max(len(word), maxlen)`

Jaya-Hoyt commented on problems/longest_substring_w-o_repeats.py:
```diff
@@ -0,0 +1,32 @@
+        word = ''
+        maxlen = 0
+        
+        while i < len(s):
```
this is O(n^2) time, right? we can do better. think about the repeated work being done with this nested loop.

Jaya-Hoyt replied:
<details>
  <summary>Hint</summary>
  
  you're building the same substrings of each longest unique string sequence
  E.g. abcdd
  first run builds abcd
  second run builds bcd
  third run builds cd

  see the wasted work?
</details>

Jaya-Hoyt commented on problems/median_two_sorted_arrays.py:
```diff
@@ -0,0 +1,36 @@
+                merged.append(nums2[j])
+                j += 1
+                
+        if j == len(nums2):
```
you don't need the if statement here and in the opposite case. your while statement should skip if it's all processed

Jaya-Hoyt commented on problems/median_two_sorted_arrays.py:
```diff
@@ -0,0 +1,36 @@
+
+from typing import List
+
+class Solution:
```
This is a nice brute force solution, but we can do better.

leetcode wants O(log (m+n)). What does that mean?
<details>
  <summary>Hint 1</summary>
  
  It means we need to cut our search space in half every time. 

  Is there a property of the input arrays that is helpful here?
  
</details>

<details>
  <summary>Hint 2</summary>
  
  The input arrays are sorted. 

  how can we find a value in a sorted array quickly?
  
</details>

<details>
  <summary>Hint 3</summary>
  
  A binary search. 
  
  Now how do we use a binary search to find a median between two arrays?
  
</details>

Jaya-Hoyt replied:
The point I want to hit home with this comment is to use time/space complexity bounds to help you narrow what algorithms and data structure are available to you. You then look at those options left to see what can solve your problem. This is a key process in optimization. find the upper bound (brute force) and identify what algorithms and data structures are available in lower bounds that could be applied to your problem space in a novel way. Being given a target big O should be a huge hint that you should never waste.

Jaya-Hoyt reviewed (changes requested):
Good stuff. the median problem is much harder than the others, but i'm interested to see how you think about the problem with the hints i gave. there are youtube videos and examples online for the solution if you're still stuck.

To reiterate from my comments
* list your big O for your main functions
    * if you write a brute force or suboptimal solution along the way, keep it around, name it <some_problem_brute_force> and list the time/space complexity. iterative improvements are an important process, and they should be shown here if you write them up.
* write the leetcode number at the front of your filenames to keep better organization of your problems.

Jaya-Hoyt commented:
I want you to write all your leetcode problems with pytest unit tests like [this](https://github.com/JustinHoyt/interview-practice/blob/master/leetcode/0001_two_sum.py). 

This lets you drag and drop your solution class in to run leetcode's tests while also letting you write your own tests first. You shouldn't be writing code then running it against their tests to see how you did. You need to create your own tests, ideally before you start coding, to formulate your best solutions.

```bash
pip3 install pytest
pip3 install pytest-watch
ptw problems/median_two_sorted_arrays.py
```

Jaya-Hoyt commented:
One more thing. I added a `__main__` check to my [example](https://github.com/JustinHoyt/interview-practice/blob/master/leetcode/0001_two_sum.py#L36) as a way to neatly be able to run your debugger against any test case while not interrupting your test watcher as it runs in the background.

If that's confusing, let me know and I'll show you how to set up your environment for easy access to a test watcher and debugger.

nwestallen committed:
adds leetcode numbers to filenames

nwestallen committed:
fixes problem links

nwestallen committed:
adds tests, comments, and imporved solution to problem 3

- pytest and ptw used to write and run test cases
- minor changes to brute force solution
- added sliding window solution
- added __pycache__ to gitignore...created when I started ptw

nwestallen committed:
adds testability to problem 2

nwestallen committed:
adds tests to problem 2

nwestallen committed:
adds testing to problem 4
